aws-lc-rs/src/pqdsa/key_pair.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+use crate::aws_lc::{EVP_PKEY_CTX_pqdsa_set_params, EVP_PKEY, EVP_PKEY_PQDSA};
+use crate::encoding::{AsDer, AsRawBytes, Pkcs8V1Der, PqdsaPrivateKeyRaw};
+use crate::error::{KeyRejected, Unspecified};
+use crate::evp_pkey::No_EVP_PKEY_CTX_consumer;
+use crate::pkcs8;
+use crate::pkcs8::{Document, Version};
+use crate::pqdsa::signature::{PqdsaSigningAlgorithm, PublicKey};
+use crate::pqdsa::validate_pqdsa_evp_key;
+use crate::ptr::LcPtr;
+use crate::signature::KeyPair;
+use core::fmt::{Debug, Formatter};
+use std::ffi::c_int;
+
+/// A PQDSA (Post-Quantum Digital Signature Algorithm) key pair, used for signing and verification.
+#[allow(clippy::module_name_repetitions)]
+pub struct PqdsaKeyPair {
+    algorithm: &'static PqdsaSigningAlgorithm,
+    evp_pkey: LcPtr<EVP_PKEY>,
+    pubkey: PublicKey,
+}
+
+#[allow(clippy::missing_fields_in_debug)]
+impl Debug for PqdsaKeyPair {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PqdsaKeyPair")
+            .field("algorithm", &self.algorithm)
+            .finish()
+    }
+}
+
+impl KeyPair for PqdsaKeyPair {
+    type PublicKey = PublicKey;
+
+    fn public_key(&self) -> &Self::PublicKey {
+        &self.pubkey
+    }
+}
+
+/// A PQDSA private key.
+pub struct PqdsaPrivateKey<'a>(pub(crate) &'a PqdsaKeyPair);
+
+impl AsDer<Pkcs8V1Der<'static>> for PqdsaPrivateKey<'_> {
+    /// Serializes the key to PKCS#8 v1 DER.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` if serialization fails.
+    fn as_der(&self) -> Result<Pkcs8V1Der<'static>, Unspecified> {
+        Ok(Pkcs8V1Der::new(
+            self.0
+                .evp_pkey
+                .marshal_rfc5208_private_key(pkcs8::Version::V1)?,
+        ))
+    }
+}
+
+impl AsRawBytes<PqdsaPrivateKeyRaw<'static>> for PqdsaPrivateKey<'_> {
+    fn as_raw_bytes(&self) -> Result<PqdsaPrivateKeyRaw<'static>, Unspecified> {
+        Ok(PqdsaPrivateKeyRaw::new(
+            self.0.evp_pkey.marshal_raw_private_key()?,
+        ))
+    }
+}
+
+impl PqdsaKeyPair {
+    /// Generates a new PQDSA key pair for the specified algorithm.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` is the key generation fails.
+    pub fn generate(algorithm: &'static PqdsaSigningAlgorithm) -> Result<Self, Unspecified> {
+        let evp_pkey = evp_key_pqdsa_generate(algorithm.0.id.nid())?;
+        let pubkey = PublicKey::from_private_evp_pkey(&evp_pkey)?;
+        Ok(Self {
+            algorithm,
+            evp_pkey,
+            pubkey,
+        })
+    }
+
+    /// Parses a PKCS#8 key from the specified bytes.
```

Comment:
Minor nit: Should we align the documentation for `from_pkcs8` and `from_raw_private_key` to be more similar? "Constructs a key pair from ... " is a great accurate description.